fs-storage/src/base_storage.rs

```diff
@@ -0,0 +1,41 @@
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::str::FromStr;
+
+use data_error::Result;
+
+pub trait BaseStorage<K, V>
+where
+    K: FromStr + Hash + Eq + Ord + Debug + Clone,
+    V: Debug + Clone,
+{
+    fn get(&self, id: &K) -> Option<V>;
+    fn set(&mut self, id: K, value: V) -> Result<()>;
+    fn remove(&mut self, id: &K) -> Result<()>;
+
+    /// Remove file at stored path
+    fn erase(&mut self) -> Result<()>;
+
+    /// Get immutable BTreeMap
+    ///
+    /// This can be used to get most immutable BtreeMap related functions for free.
+    fn as_ref(&self) -> &BTreeMap<K, V>;
```

Comment:
I am not sure if this is truly necessary.
If `FileStorage` and `FolderStorage` directly manage their own internal storage (e.g., a `BTreeMap`), they might not need to expose it directly.

Reply:
cc @twitu

Reply:
By exposing it directly we can provide the user all the immutable operations for free. Otherwise if a user wants to use the btree as an iterator we will have to add in a wrapper. Similarly for each extra function that needs to be supported `ark-rust` will have to add a wrapper.

In any case, I was going for something along these lines.

```
pub trait BaseStorage<K, V>: AsRef<BTreeMap<K, V>>
where
  ...
```

This way anything that implements BaseStorage must also be referencable as a BtreeMap which fits are notion that BaseStorage is a virtualization of a key value mapping stored on disk.

[trait inheritance](https://users.rust-lang.org/t/extending-traits/1802)
[super trait](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-supertraits-to-require-one-traits-functionality-within-another-trait)

Reply:
Got it. Thanks for the clarification.